docker/indexer/stages/preparation/preparation.go

```diff
@@ -0,0 +1,294 @@
+type Stage struct {
+	Checker Checker
+	RepoHdl *storage.BucketHandle
+}
+
+// Run runs the stage and outputs Result data types to the results channel. 
+func (s *Stage) Run(ctx context.Context, cfgs []*pb.Repository, results chan *Result) error {
+	var err error
+	wErr := make(chan error, workers)
+	wg := sync.WaitGroup{}
+	routineCtr := 0
+	wCtx, wCancel := context.WithCancel(ctx)
+	for _, repoCfg := range cfgs {
+		if routineCtr >= workers {
+			wg.Wait()
+			routineCtr = 0
+		}
+		go func(ctx context.Context, repoCfg *pb.Repository, results chan *Result, errCh chan error) {
+			defer wg.Done()
+
+			var err error
+			select {
+			case <-ctx.Done():
+				errCh <- context.Canceled
+				return
+			default:
+			}
+			log.Infof("received config for %s", repoCfg.Name)
+			switch repoCfg.Type {
+			case pb.RepositoryType_GIT:
+				err = s.processGit(ctx, repoCfg, results)
+			default:
+				errCh <- fmt.Errorf("unsupported repository type %s", repoCfg.Type.String())
+			}
+			if err != nil {
+				wErr <- err
+			}
+		}(wCtx, repoCfg, results, wErr)
+
+		wg.Add(1)
+		routineCtr++
+		select {
+		case err = <-wErr:
+			wCancel()
+		default:
+		}
+		if err != nil {
+			break
+		}
+	}
+	wg.Wait()
+	wCancel()
+	return err
+}
+
+func (s *Stage) objectExists(ctx context.Context, name string) bool {
+	objItr := s.RepoHdl.Objects(ctx, &storage.Query{Prefix: name + shared.TarExt})
+	_, err := objItr.Next()
+	return err == nil
+
+}
+
+func (s *Stage) processGit(ctx context.Context, repoCfg *pb.Repository, results chan *Result) error {
+	var (
+		err     error
+		repo    *git.Repository
+		repoDir string
+	)
+	if !s.objectExists(ctx, repoCfg.Name) {
+		repo, repoDir, err = s.cloneGitRepo(ctx, repoCfg.Name, repoCfg.Address)
+	} else {
+		repo, repoDir, err = s.updateGitRepo(ctx, repoCfg.Name)
+	}
+	if repoDir != "" {
+		defer func() {
+			if err := os.RemoveAll(repoDir); err != nil {
+				log.Errorf("failed to remove local repo: %v", err)
+			}
+		}()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	comItr, err := repo.CommitObjects()
+	if err != nil {
+		return err
+	}
+	allCommits := make(map[plumbing.Hash]*object.Commit)
+	comItr.ForEach(func(c *object.Commit) error {
+		allCommits[c.Hash] = c
+		return nil
+	})
+
+	commitTracker := make(map[plumbing.Hash]bool)
+	// repoInfo is used as the iterator function to create RepositoryInformation structs.
+	repoInfo := func(ref *plumbing.Reference) error {
+		found, err := s.Checker.Exists(ctx, repoCfg.Name, ref.Hash())
+		if err != nil {
+			return err
+		}
+		if found {
+			return nil
+		}
+
+		var when time.Time
+		if c, ok := allCommits[ref.Hash()]; ok {
+			when = c.Author.When
+		}
+		versRE, err := regexp.Compile(repoCfg.VersionRegex)
+		if err != nil {
+			return err
+		}
+		results <- &Result{
+			Name:        repoCfg.Name,
+			BaseCPE:     repoCfg.BaseCpe,
+			VersionRE:   versRE,
+			TagMatching: true,
+			CheckoutOptions: &git.CheckoutOptions{
+				Branch: ref.Name(),
+				Force:  true,
```

Comment:
What will this field be used for?

Reply:
This is used in processing to check out the corresponding branch/commit.